### ISSUE TITLE
char in string is slow for short strings

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#43980]
From [discourse](https://discourse.julialang.org/t/rust-julia-comparison-post/75403/21?u=stevengj)
```jl
s = "chalet"
v = Vector{Char}(s)
@btime in('t', $(Ref(s))[]) # 17.077 ns (0 allocations: 0 bytes)
@btime in('t', $(Ref(v))[]) # 5.514 ns (0 allocations: 0 bytes)
```
though for a longer string it is faster:
```jl
s = Random.randstring(1000) * '='
v = Vector{Char}(s)
@btime in('t', $(Ref(s))[]) # 18.601 ns (0 allocations: 0 bytes)
@btime in('t', $(Ref(v))[]) # 46.032 ns (0 allocations: 0 bytes)
```
I don't see any inherent reason why this should be the case, especially when searching for ASCII chars.

**Comment by KristofferC** [comment]
@jakobnissen, is this fixed now?

**Comment by jakobnissen** [comment]
No, unfortunately, it's still slower. On my computer, 4.3 ns for the array versus 6.5 ns for the string. Here are some reasons:
* Searching a string must throw if the starting index is invalid
* For very short strings, calling `memchr` might have a few nanoseconds overhead, making a manual search faster for short strings
* Since the string search function is more complex, in the generic definition `findfirst(f, A) = findnext(f, A, first(keys(A)))`, the call to `findnext` doesn't inline, which disables a bunch of extra optimisations.

Some ways we can address this:
1. Add a check in `_search` that checks if the string is sufficiently short, and if so, use a manual loop instead of calling into memchr.
2. Add a callsite `@inline` in the generic definition of `findfirst` - after all, it does nothing other than calling `findnext`. The problem there is of course if computing `first(keys(A))` is unexpectedly expensive, we might not want to also inline `findnext`.

I can give point 1 a go and see how the extra branch affects timing, and I'm not sure about how wise point 2 is.
